File: dotnet/Lexy/Lexy.Compiler/Language/Expressions/ParseExpressionResult.cs
namespace Lexy.Compiler.Language.Expressions
{
    public class ParseExpressionResult
    {
        public Expression Expression { get; }
        public string ErrorMessage { get; }
        public ParseExpressionStatus Status { get; }

        private ParseExpressionResult(ParseExpressionStatus status, string errorMessage)
        {
            Status = status;
            ErrorMessage = errorMessage;
            Status = status;
        }

        private ParseExpressionResult(ParseExpressionStatus status, Expression expression = null)
        {
            Expression = expression;
            Status = status;
        }

        public static ParseExpressionResult Invalid<T>(string error)
        {
            return new ParseExpressionResult(ParseExpressionStatus.Failed, $"({typeof(T).Name}) {error}");
        }

        public static ParseExpressionResult Success(Expression expression)
        {
            return new ParseExpressionResult(ParseExpressionStatus.Success, expression);
        }
    }
}

File: typescipt/existing/src/language/expressions/parseExpressionResult.ts
namespace Lexy.Compiler.Language.Expressions;

public sealed class ParseExpressionResult : ParseResult<Expression>
{
   private ParseExpressionResult(Expression result) : base(result)
   {
   }

   private ParseExpressionResult(bool success, string errorMessage) : base(success, errorMessage)
   {
   }

   public static ParseExpressionResult Invalid<T>(string error)
   {
     return new ParseExpressionResult(false, $"({typeof(T).Name}) {error}");
   }

   public static ParseExpressionResult Success(Expression expression)
   {
     return new ParseExpressionResult(expression);
   }
}
